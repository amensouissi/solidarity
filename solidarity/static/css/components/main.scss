.container {
  background-color: $blue;
}

.title {
  color: gray;
}
