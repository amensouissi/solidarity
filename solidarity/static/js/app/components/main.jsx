// @flow
import * as React from 'react';
import { graphql } from 'react-apollo';
import { withStyles } from '@material-ui/core/styles';
import AppBar from '@material-ui/core/AppBar';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import ThumbUp from '@material-ui/icons/ThumbUp';
import Typography from '@material-ui/core/Typography';
import { Link } from 'react-router';

import Pages from '../graphql/queries/Pages.graphql';

type Page = {
  id: string,
  title: string,
  body: string
};

type Props = {
  classes: Object,
  data: {
    loading: boolean,
    pages: Array<Page>
  },
  children: React.Node
};

type State = {
  value: number,
  hasError: boolean
};

type TabContainerProps = {
  children: Object
};

function TabContainer({ children }: TabContainerProps) {
  return (
    <Typography component="div" style={{ padding: 8 * 3 }}>
      {children}
    </Typography>
  );
}

const styles = {
  root: {
    flexGrow: 1,
    width: '100%'
  }
};

export class DumbMain extends React.Component<Props, State> {
  state = {
    value: 0,
    hasError: false
  };

  componentDidCatch() {
    this.setState({ hasError: true });
  }

  handleChange = (event: SyntheticEvent<HTMLButtonElement>, value: number) => {
    this.setState({ value: value });
  };

  render() {
    const { data, classes, children } = this.props;
    const { value, hasError } = this.state;
    if (data.loading) return 'Loading...';
    const currentPage = data.pages[value];
    return (
      <div className={classes.root}>
        <AppBar position="static" color="default">
          <Tabs
            value={value}
            onChange={this.handleChange}
            scrollable
            scrollButtons="on"
            indicatorColor="primary"
            textColor="primary"
          >
            {data.pages.map((page: Page) => {
              return <Tab key={page.id} label={page.title} icon={<ThumbUp />} />;
            })}
          </Tabs>
        </AppBar>
        {currentPage ? (
          <TabContainer>
            <Link to="/page">
              <div dangerouslySetInnerHTML={{ __html: currentPage.body }} />
            </Link>
          </TabContainer>
        ) : null}
        {hasError ? 'Oups an error!' : children}
      </div>
    );
  }
}

// TODO use Query component instead of graphql
export default withStyles(styles)(graphql(Pages)(DumbMain));