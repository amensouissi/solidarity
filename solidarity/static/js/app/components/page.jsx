// @flow
import * as React from 'react';

type Props = {
  start?: number
};

type State = {
  /* the counter */
  counter: number,
  timeout: ?TimeoutID
};

class Page extends React.Component<Props, State> {
  static defaultProps = {
    start: 0
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      // $FlowFixMe we have a default value for the start prop
      counter: props.start,
      timeout: null
    };
  }

  componentDidMount() {
    this.start();
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return nextState.counter !== 10;
  }

  componentDidUpdate() {
    const { counter } = this.state;
    if (counter > 6) throw new Error('Error');
  }

  componentWillUnmout() {
    const { timeout } = this.state;
    if (timeout) clearTimeout(timeout);
  }

  start = () => {
    this.setState({ timeout: setTimeout(this.clic, 1000) });
  };

  clic = () => {
    this.setState((state) => {
      return { ...state, counter: state.counter + 1 };
    }, this.start);
  };

  render() {
    const { counter } = this.state;
    return <strong>{counter}</strong>;
  }
}

export default Page;