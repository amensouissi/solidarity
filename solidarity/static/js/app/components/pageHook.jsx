/* eslint-disable react/destructuring-assignment */
import React, { useState, useEffect } from 'react';

function Page(props) {
  const [counter, setCount] = useState(props.start);

  function clic() {
    setCount(counter + 1);
  }

  useEffect(() => {
    const timeout = setTimeout(clic, 1000);
    return () => {
      clearTimeout(timeout);
    };
  });

  return <strong>{counter}</strong>;
}

Page.defaultProps = { start: 0 };

export default Page;