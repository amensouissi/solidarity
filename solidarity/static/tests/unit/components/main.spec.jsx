import React from 'react';
import renderer from 'react-test-renderer';

import { DumbMain } from '../../../js/app/components/main';

describe('Main component', () => {
  it('should render the main component', () => {
    const props = {
      data: { loading: false, comment: 'FooBar' }
    };
    const component = renderer.create(<DumbMain {...props} />);
    const result = component.toJSON();
    expect(result).toMatchSnapshot();
  });
  it('should render the main component (loading)', () => {
    const props = {
      data: { loading: true }
    };
    const component = renderer.create(<DumbMain {...props} />);
    const result = component.toJSON();
    expect(result).toMatchSnapshot();
  });
});