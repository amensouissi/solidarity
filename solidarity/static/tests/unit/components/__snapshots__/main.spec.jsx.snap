// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main component should render the main component (loading) 1`] = `"Loading..."`;

exports[`Main component should render the main component 1`] = `
<div>
  <WithStyles(AppBar)
    color="default"
    position="static"
  >
    <WithStyles(Tabs)
      indicatorColor="primary"
      onChange={[Function]}
      scrollButtons="on"
      scrollable={true}
      textColor="primary"
      value={0}
    >
      <WithStyles(Tab)
        icon={<pure(ThumbUpIcon) />}
        label="Foo"
      />
      <WithStyles(Tab)
        icon={<pure(ThumbUpIcon) />}
        label="Bar"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <TabContainer>
    <Link
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/page"
    >
      <div
        dangerouslySetInnerHTML={
          Object {
            "__html": "Foo body",
          }
        }
      />
    </Link>
  </TabContainer>
</div>
`;
